hashmap/src/main.rs

```diff
@@ -0,0 +1,273 @@
+}
+
+#[derive(thiserror::Error, Debug)]
+pub enum MeasurementStreamError {
+    #[error("Unexpected time stamp within a group")]
+    UnexpectedTimestamp,
+
+    #[error("Unexpected end of data")]
+    UnexpectedEndOfData,
+
+    #[error("Unexpected end of group")]
+    UnexpectedEndOfGroup,
+
+    #[error("Unexpected start of group")]
+    UnexpectedStartOfGroup,
+}
+
+impl ThinEdgeJsonSerializer {
+    pub fn new() -> Self {
+        ThinEdgeJsonSerializer {
+            buffer: Vec::new(),
+            is_within_group: false,
+            needs_separator: false,
+        }
+    }
+    fn tej_serailized(&mut self) -> Vec<u8> {
+        self.buffer.clone()
+    }
+}
+
+impl MeasurementConsumer for ThinEdgeJsonSerializer {
+    type Error = ThinEdgeJsonSerializationError;
+    type Data = Vec<u8>;
+    fn start(&mut self) -> Result<(), Self::Error> {
+        self.buffer.push(b'{');
+        self.needs_separator = false;
+        Ok(())
+    }
+
+    fn end(&mut self) -> Result<(), Self::Error> {
+        if self.is_within_group {
+            return Err(MeasurementStreamError::UnexpectedEndOfData.into());
+        }
+
+        self.buffer.push(b'}');
+        //println!("vector {:#?}", std::str::from_utf8(&self.buffer));
+        Ok(())
+    }
+
+    fn timestamp(&mut self, value: String) -> Result<(), Self::Error> {
+        if self.is_within_group {
+            return Err(MeasurementStreamError::UnexpectedTimestamp.into());
+        }
+
+        if self.needs_separator {
+            self.buffer.push(b',');
+        }
+        self.buffer
+            .write_fmt(format_args!("\"time\":\"{}\"", value))?;
+        self.needs_separator = true;
+        Ok(())
+    }
+
+    fn measurement(
+        &mut self,
+        _grpname: Option<&str>,
+        name: &str,
+        value: f64,
+    ) -> Result<(), Self::Error> {
+        if self.needs_separator {
+            self.buffer.push(b',');
+        }
+        self.buffer
+            .write_fmt(format_args!("\"{}\":{}", name, value))?;
+        self.needs_separator = true;
+        Ok(())
+    }
+
+    fn start_group(&mut self, group: &str) -> Result<(), Self::Error> {
+        if self.is_within_group {
+            return Err(MeasurementStreamError::UnexpectedStartOfGroup.into());
+        }
+
+        if self.needs_separator {
+            self.buffer.push(b',');
+        }
+        self.buffer.write_fmt(format_args!("\"{}\":{{", group))?;
+        self.needs_separator = false;
+        Ok(())
+    }
+
+    fn end_group(&mut self) -> Result<(), Self::Error> {
+        if !self.is_within_group {
+            return Err(MeasurementStreamError::UnexpectedEndOfGroup.into());
+        }
+
+        self.buffer.push(b'}');
+        self.needs_separator = true;
+        Ok(())
+    }
+}
+
+fn main() {
+    //Produce the TEJ from raw data
+    let mut tej_producer = MeasurementMap::new();
+    tej_producer.timestamp("4-20-2020".into()).unwrap();
```

Comment:
These `unwrap` calls are a bit ugly and takes away the beauty of chaining from the APIs. Any way to avoid that?

Reply:
The unwraps there just because its PoC.